silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
+	// on disk.
+	msil := s.toMeshSilence(sil)
+	if err := s.checkLimits(msil, updateExisting); err != nil {
+		return fmt.Errorf("exceeds limits: %w", err)
+	}
+
+	// expireExisting is true iff this is an existing silence, and the update
+	// affects how alerts are matched or the time window in which the silence
+	// is active (excluding extensions). If existing and sil reference the
+	// same memory, then the operation will fail because the new ID does not
+	// exist. This happens when the caller does not use Clone().
+	expireExisting := ok && !updateExisting && getState(existing, now) != types.SilenceStateExpired
+	if expireExisting {
+		if err := s.expire(existing.Id); err != nil {
+			return fmt.Errorf("failed to expire silence: %w", err)
+		}
 	}
 
-	// Check the limit unless the silence has been expired. This is to avoid
-	// situations where silences cannot be expired after the limit has been
-	// reduced.
+	return s.mergeAndBroadcast(msil, now)
+}
+
+func (s *Silences) checkLimits(msil *pb.MeshSilence, updateExisting bool) error {
 	if s.limits.MaxSilenceSizeBytes != nil {
 		n := msil.Size()
-		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m && sil.EndsAt.After(now) {
-			return fmt.Errorf("silence exceeded maximum size: %d bytes (limit: %d bytes)", n, m)
+		if m := s.limits.MaxSilenceSizeBytes(); m > 0 && n > m {
+			return fmt.Errorf("maximum size: %d bytes (limit: %d bytes)", n, m)
+		}
+	}
+	if s.limits.MaxSilences != nil && !updateExisting {
+		if m := s.limits.MaxSilences(); m > 0 && len(s.st)+1 > m {
+			return fmt.Errorf("maximum number of silences: %d (limit: %d)", len(s.st), m)
 		}
 	}
+	return nil
+}
 
+func (s *Silences) getSilence(id string) (*pb.Silence, bool) {
+	msil, ok := s.st[id]
+	if !ok {
+		return nil, false
+	}
+	return msil.Silence, true
+}
+
+func (s *Silences) mergeAndBroadcast(msil *pb.MeshSilence, now time.Time) error {
+	b, err := marshalMeshSilence(msil)
+	if err != nil {
+		return fmt.Errorf("failed to marshal silence: %w", err)
+	}
 	if s.st.merge(msil, now) {
 		s.version++
 	}
 	s.broadcast(b)
-
 	return nil
 }
 
-// Set the specified silence. If a silence with the ID already exists and the modification
-// modifies history, the old silence gets expired and a new one is created.
-func (s *Silences) Set(sil *pb.Silence) error {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+func (s *Silences) toMeshSilence(sil *pb.Silence) *pb.MeshSilence {
+	return &pb.MeshSilence{
+		Silence:   sil,
+		ExpiresAt: sil.EndsAt.Add(s.retention),
+	}
+}
 
-	now := s.nowUTC()
-	prev, ok := s.getSilence(sil.Id)
-	if sil.Id != "" && !ok {
+// Expire the silence with the given ID immediately.
+// It is idempotent, nil is returned if the silence already expired before it is GC'd.
+// If the silence is not found an error is returned.
+func (s *Silences) expire(id string) error {
+	sil, ok := s.getSilence(id)
+	if !ok {
 		return ErrNotFound
 	}
+	sil = shallowClone(sil)
```

Comment:
Renamed `cloneSilence` to `shallowClone` since we now have a deep copy function called `Clone`.